src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,407 @@
+			if (getErrorChannel() != null) {
+				AttributeAccessor attributes = ErrorMessageUtils.getAttributeAccessor(message, null);
+				attributesHolder.set(attributes);
+				attributes.setAttribute(AwsHeaders.RAW_RECORD, record);
+			}
+		}
+
+		private AbstractIntegrationMessageBuilder<Object> prepareMessageForRecord(KinesisClientRecord record) {
+			ByteBuffer data = record.data();
+			byte[] dataArray = new byte[data.remaining()];
+			Object payload = dataArray;
+			data.get(dataArray);
+			Message<?> messageToUse = null;
+
+			if (KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper != null) {
+				try {
+					messageToUse = KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper.toMessage((byte[]) payload);
+
+					payload = messageToUse.getPayload();
+				}
+				catch (Exception e) {
+					logger.warn("Could not parse embedded headers. Remain payload untouched.", e);
+				}
+			}
+
+			AbstractIntegrationMessageBuilder<Object> messageBuilder = getMessageBuilderFactory().withPayload(payload)
+					.setHeader(AwsHeaders.RECEIVED_PARTITION_KEY, record.partitionKey())
+					.setHeader(AwsHeaders.RECEIVED_SEQUENCE_NUMBER, record.sequenceNumber())
+					.setHeader(AwsHeaders.RECEIVED_STREAM, KclMessageDrivenChannelAdapter.this.stream)
+					.setHeader(AwsHeaders.SHARD, this.shardId);
+
+			if (messageToUse != null) {
+				messageBuilder.copyHeadersIfAbsent(messageToUse.getHeaders());
+			}
+
+			return messageBuilder;
+		}
+
+		/**
+		 * Checkpoint with retries.
+		 *
+		 * @param checkpointer checkpointer
+		 */
+		private void checkpoint(RecordProcessorCheckpointer checkpointer) {
+			if (logger.isInfoEnabled()) {
+				logger.info("Checkpointing shard " + shardId);
+			}
+			try {
+				checkpointer.checkpoint();
+			}
+			catch (ShutdownException se) {
+				// Ignore checkpoint if the processor instance has been shutdown (fail over).
+				logger.info("Caught shutdown exception, skipping checkpoint.", se);
+			}
+			catch (ThrottlingException e) {
+				if (logger.isInfoEnabled()) {
+					logger.info("Transient issue when checkpointing", e);
+				}
+			}
+			catch (InvalidStateException e) {
+				// This indicates an issue with the DynamoDB table (check for table, provisioned
+				// IOPS).
+				logger.error("Cannot save checkpoint to the DynamoDB table used by the Amazon Kinesis Client Library.",
+						e);
+			}
+		}
+
+		@Override
+		public void processRecords(ProcessRecordsInput processRecordsInput) {
+			List<KinesisClientRecord> records = processRecordsInput.records();
+			if (logger.isDebugEnabled()) {
+				logger.debug("Processing " + records.size() + " records from " + this.shardId);
+			}
+
+			// Process records and perform all exception handling.
+			processRecordsWithRetries(records);
+
+			// Checkpoint once every checkpoint interval.
+			if (System.currentTimeMillis() > nextCheckpointTimeInMillis) {
+				checkpoint(processRecordsInput.checkpointer());
```

Comment:
Why do we checkpoint for the whole batch?
I believe in case of single record to send we need to checkpoint per record.

Also I believe that we need to have `manual` mode when end-user will decide by himself when and how to checkpoint.

See similar logic in the `KinesisMessageDrivenChannelAdapter`.

We may move this concern to the separate issue if you are good to proceed with it afterwards.

Reply:
I'm not sure, but I believe we did this because we were concerned about performance issues. For one use case we can have more than 10000 records per second, so the performance impact for checkpointing on each record could be a problem.
Maybe we could implement 3 modes ? auto (checkpoint at each record), manual (as in the current kinesis adapter) and periodic (what we implemented)
We can take care of it later if it's ok with you to do it in a separate issue.

Reply:
Sounds good.

I think we need to have a `batch` checkpoint as well. 
See `CheckpointMode`.

I don't mind about adding a new `periodic` one based on the `checkpointsInterval` option, which could do be introduced for the `KinesisMessageDrivenChannelAdapter` as well.
In separate issue. Agreed.